File: 5-Problem-A-Day/272. Closest Binary Search Tree Value II.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<Integer> closestKValues(TreeNode root, double target, int k) {
        LinkedList<Integer> ans = new LinkedList<>();
        collect(root, target, k, ans);
        return ans;
    }

    private void collect(TreeNode root, double target, int k, LinkedList<Integer> ans) {
        if (root == null)
            return;
        collect(root.left, target, k, ans);
        if (ans.size() == k) {
            if (Math.abs(target - root.val) < Math.abs(target - ans.peekFirst()))
                ans.removeFirst();
            else
                return;
        }
        ans.add(root.val);
        collect(root.right, target, k, ans);
    }
}

File: 5-Problem-A-Day/272. Closest Binary Search Tree Value II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:    
    vector<int> closestKValues(TreeNode* root, double target, int k) {        
        vector<int> vals, res;
        inorder(root, vals);
        int index = lower_bound(vals.begin(), vals.end(), target) - vals.begin();   
        index = min(index, (int)vals.size() - 1);
        int left = index - 1, right = index;           
        while (res.size() < k) {            
            if (left < 0 || (right < vals.size() && abs(vals[left] - target) > abs(vals[right]) - target)) {
                res.push_back(vals[right]);
                right++;
            } else {
                res.push_back(vals[left]);
                left--;
            }
        }
        
        return res;
    }
    
    void inorder(TreeNode* root, vector<int>& vals) {
        if (root == NULL) return;
        inorder(root->left, vals);
        vals.push_back(root->val);
        inorder(root->right, vals);
    }
    
    
    
};